### PR TITLE
MAT-5237 Draft a Measure

### PR DESCRIPTION
**Pull request by bwelch8105**
Created automated tests for MAT-5237 and necessary updates to existing support files.

### CODE REVIEW

**Review by cvasile001 — CHANGES_REQUESTED** [review]
I think you need to add a test case for checking to see you that you cannot draft if a draft already exists for that measure family

Also, not sure the changes you are making to RunandExecuteTestCase. Just formatting changes? anything to do with drafting?

File: cypress/Shared/MeasuresPage.ts [hunk 1]
```diff
@@ -14,6 +14,11 @@ export class MeasuresPage {
     public static readonly paginationLimitEquals25 = '[data-value="25"]'
 
     //Measure Version
+    public static readonly versionMeasuresRadioButton = '[name="type"]'
+    public static readonly VersionDraftMsgs = '[data-testid="success-toast"]'
```
[the remark this answers is not in the document]
Comment (bwelch8105) [reply cypress/Shared/MeasuresPage.ts:18]: Added tests and updated other tests.

There may be some formatting updates with this PR due to my linter

File: cypress/e2e/WebInterface/Measure/VersionDraftMeasure/DraftMeasure.cy.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+        MeasuresPage.clickDraftforCreatedMeasure()
+        cy.get(MeasuresPage.updateDraftedMeasuresTextBox).clear().type(updatedMeasuresPageName + '1')
+        cy.get(MeasuresPage.createDraftContinueBtn).click()
+        cy.get(MeasuresPage.VersionDraftErrMsgs).should('contains.text', 'Can not create a draft for the measure ')
+        cy.get(MeasuresPage.VersionDraftErrMsgs).should('contains.text', '. Only one draft is permitted per measure.')
+    })
+
+    it('Verify the Measure updates are restricted after Version is created', () => {
+
```
[the remark this answers is not in the document]
Comment (bwelch8105) [reply cypress/e2e/WebInterface/Measure/VersionDraftMeasure/DraftMeasure.cy.ts:129]: You are correct -- this test already existed so I removed it from this PR.

File: cypress/Shared/MeasuresPage.ts [hunk 3]
```diff
@@ -28,25 +33,24 @@ export class MeasuresPage {
         }
 
         cy.readFile(filePath).should('exist').then((fileContents) => {
-            Utilities.waitForElementVisible('[data-testid=edit-measure-' + fileContents + ']', 30000)
```
[the remark this answers is not in the document]
Comment (bwelch8105) [reply cypress/Shared/MeasuresPage.ts:31]: I have reverted to using the _old_ ClickEditforCreatedMeasure function and have commented out the newer stuff until the flag is removed. I also skipped the tests that uses the newer way.

File: cypress/Shared/MeasuresPage.ts [hunk 4]
```diff
@@ -14,11 +14,21 @@ export class MeasuresPage {
...
+    public static readonly versionMeasuresRadioButton = '[name="type"]'
+    public static readonly VersionDraftMsgs = '[data-testid="success-toast"]'
+    public static readonly VersionDraftErrMsgs = '[data-testid="error-toast"]'
+    public static readonly updateDraftedMeasuresTextBox = '[data-testid="measure-name-input"]'
+    public static readonly createDraftContinueBtn = '[data-testid="create-draft-continue-button"]'
     public static readonly measureVersionMajor = '[data-testid="radio-button-group"] > :nth-child(1)'
     public static readonly measureVersionContinueBtn = '[data-testid="create-version-continue-button"]'
     public static readonly measureVersionSuccessMsg = '[data-testid="success-toast"]'
     public static readonly measureVersioningErrorMsg = '[data-testid="error-toast"]'
+    public static readonly measureSelectBtn = '#measure-landing > div > div > div > div:nth-child(1) > div > div > div > div:nth-child(1) > table > tbody > tr > td:nth-child(4) > button'
```
Comment (cvasile001) [inline cypress/Shared/MeasuresPage.ts:26]: which button is this? lets see if DEV can get a testid for it

File: cypress/Shared/MeasuresPage.ts [hunk 5]
```diff
@@ -14,11 +14,21 @@ export class MeasuresPage {
...
+    public static readonly measureSelectContainer = '[class="btn-container" ]'
 
+    //<div class="btn-container"><button data-testid="view-measure-63d40acb1632e36a0f9f9653">View</button><button data-testid="create-version-measure-63d40acb1632e36a0f9f9653">Version</button></div>
+
```
Comment (cvasile001) [inline cypress/Shared/MeasuresPage.ts:30]: not sure if we need these comments

File: cypress/Shared/MeasuresPage.ts [hunk 6]
```diff
@@ -27,26 +37,62 @@ export class MeasuresPage {
...
+                }) */
         cy.readFile(filePath).should('exist').then((fileContents) => {
             Utilities.waitForElementVisible('[data-testid=edit-measure-' + fileContents + ']', 30000)
             cy.get('[data-testid=edit-measure-' + fileContents + ']').should('be.visible')
             Utilities.waitForElementEnabled('[data-testid=edit-measure-' + fileContents + ']', 30000)
             cy.get('[data-testid=edit-measure-' + fileContents + ']').should('be.enabled')
             cy.get('[data-testid=edit-measure-' + fileContents + ']').click()
+        })
+    }
+    public static clickEditforCreatedDraft(secondDraft?: boolean): void {
```
Comment (cvasile001) [inline cypress/Shared/MeasuresPage.ts:62]: I don't think that we need this, but maybe there is a reason I am not thinking of?

File: cypress/Shared/MeasuresPage.ts [hunk 7]
```diff
@@ -79,8 +125,22 @@ export class MeasuresPage {
...
+
+        cy.readFile('cypress/fixtures/measureId').should('exist').then((fileContents) => {
+            Utilities.waitForElementVisible('[data-testid=measure-action-' + fileContents + ']', 30000)
+            cy.get('[data-testid=measure-action-' + fileContents + ']').should('be.visible')
+            Utilities.waitForElementEnabled('[data-testid=measure-action-' + fileContents + ']', 30000)
+            cy.get('[data-testid=measure-action-' + fileContents + ']').should('be.enabled')
+            cy.get('[data-testid=measure-action-' + fileContents + ']').click()
+            cy.get('[data-testid=draft-measure-' + fileContents + ']').click()
+
+            //<button data-testid="draft-measure-63d2951d5a023e64c3f4d77a">Draft</button>
```
Comment (cvasile001) [inline cypress/Shared/MeasuresPage.ts:143]: don't think we need this comment

File: cypress/e2e/WebInterface/Measure/VersionDraftMeasure/DraftMeasure.cy.ts [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+let MeasuresPageOne = ''
+let updatedMeasuresPageName = ''
+let randValue = (Math.floor((Math.random() * 1000) + 1))
+let newMeasureName = ''
+let newCqlLibraryName = ''
+let ratioMeasureCQL = MeasureCQL.ICFCleanTest_CQL
+let cohortMeasureCQL = MeasureCQL.CQL_For_Cohort
+
+describe('Draft and Version Validations -- add and cannot create draft of a draft that already exists tests', () => {
```
Comment (cvasile001) [inline cypress/e2e/WebInterface/Measure/VersionDraftMeasure/DraftMeasure.cy.ts:20]: lets add some skips to these test cases, you can do it at the describe. Maybe put a comment about which FLAG we need to be removed for these test cases to be active

File: cypress/e2e/WebInterface/Test Cases/Validations/RunAndExecuteTestCaseButtonValidations.cy.ts [hunk 9]
```diff
@@ -1575,281 +1575,281 @@ describe('Verify "Run Test Cases" results based on missing/empty group populatio
     it('Can "Run Test Cases" on test case list page after a group has its population basis or scoring value changed' +
```
Comment (cvasile001) [inline cypress/e2e/WebInterface/Test Cases/Validations/RunAndExecuteTestCaseButtonValidations.cy.ts:1575]: is this just a reformat for this file?

**Review by cvasile001 — CHANGES_REQUESTED** [review]
added some questions and comments

File: cypress/Shared/MeasuresPage.ts (hunk 5 above)
Comment (bwelch8105) [reply cypress/Shared/MeasuresPage.ts:30]: removed all un-necessary comments from the MeasurePage file

File: cypress/Shared/MeasuresPage.ts (hunk 6 above)
Comment (bwelch8105) [reply cypress/Shared/MeasuresPage.ts:62]: Removed this function

File: cypress/Shared/MeasuresPage.ts (hunk 7 above)
Comment (bwelch8105) [reply cypress/Shared/MeasuresPage.ts:143]: Removed all un-necessary comments from the MeasurePage file

File: cypress/e2e/WebInterface/Measure/VersionDraftMeasure/DraftMeasure.cy.ts (hunk 8 above)
Comment (bwelch8105) [reply cypress/e2e/WebInterface/Measure/VersionDraftMeasure/DraftMeasure.cy.ts:20]: skipped all new tests and the version flag's name is measureVersioning -- I confirmed this with Prateek

File: cypress/e2e/WebInterface/Test Cases/Validations/RunAndExecuteTestCaseButtonValidations.cy.ts (hunk 9 above)
Comment (bwelch8105) [reply cypress/e2e/WebInterface/Test Cases/Validations/RunAndExecuteTestCaseButtonValidations.cy.ts:1575]: I can confirm that all of the changes seen in the RunAndExecuteTestCaseButtonValidations are just format changes. This was confirmed by switching to the master branch, pulling the latest, and, then, copying the contents out of the RunAndExecuteTestCaseButtonValidations file. Then, I switched back my branch (branch that this PR is on) and pasted the contents into the file on this branch and, then, saved. My linter only makes formatting changes when I save. So, considering that I am not pushing anything new with the file with my last push / update to this PR, the changes to this file were only formatting related.

**Review by cvasile001 — APPROVED** [review]
approved

**Pull request merged**